File: src/main/java/com/uml/generator/model/Span.java
package com.uml.generator.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Represents a Jaeger span.
 * A span represents a single operation within a distributed trace.
 */
public class Span {

    @JsonProperty("traceID")
    private String traceID;

    @JsonProperty("spanID")
    private String spanID;

    @JsonProperty("operationName")
    private String operationName;

    @JsonProperty("references")
    private List<Reference> references;

    @JsonProperty("startTime")
    private long startTime;

    @JsonProperty("duration")
    private long duration;

    @JsonProperty("tags")
    private List<Map<String, Object>> tags;

    @JsonProperty("logs")
    private List<Map<String, Object>> logs;

    @JsonProperty("processID")
    private String processID;

    @JsonProperty("warnings")
    private List<String> warnings;

    public Span() {
        this.references = new ArrayList<>();
        this.tags = new ArrayList<>();
        this.logs = new ArrayList<>();
    }

    public String getTraceID() {
        return traceID;
    }

    public void setTraceID(String traceID) {
        this.traceID = traceID;
    }

    public String getSpanID() {
        return spanID;
    }

    public void setSpanID(String spanID) {
        this.spanID = spanID;
    }

    public String getOperationName() {
        return operationName;
    }

    public void setOperationName(String operationName) {
        this.operationName = operationName;
    }

    public List<Reference> getReferences() {
        return references;
    }

    public void setReferences(List<Reference> references) {
        this.references = references;
    }

    public long getStartTime() {
        return startTime;
    }

    public void setStartTime(long startTime) {
        this.startTime = startTime;
    }

    public long getDuration() {
        return duration;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public List<Map<String, Object>> getTags() {
        return tags;
    }

    public void setTags(List<Map<String, Object>> tags) {
        this.tags = tags;
    }

    public List<Map<String, Object>> getLogs() {
        return logs;
    }

    public void setLogs(List<Map<String, Object>> logs) {
        this.logs = logs;
    }

    public String getProcessID() {
        return processID;
    }

    public void setProcessID(String processID) {
        this.processID = processID;
    }

    public List<String> getWarnings() {
        return warnings;
    }

    public void setWarnings(List<String> warnings) {
        this.warnings = warnings;
    }

    /**
     * Gets a tag value by key.
     * 
     * @param key the tag key
     * @return the tag value or null if not found
     */
    public Object getTagValue(String key) {
        if (tags == null)
            return null;

        for (Map<String, Object> tag : tags) {
            if (key.equals(tag.get("key"))) {
                return tag.get("value");
            }
        }
        return null;
    }

    /**
     * Gets a tag value by key as a String.
     * 
     * @param key the tag key
     * @return the tag value as String or null if not found
     */
    public String getTagValueAsString(String key) {
        Object value = getTagValue(key);
        return value != null ? value.toString() : null;
    }

    /**
     * Gets the parent span ID if this span has a CHILD_OF reference.
     * 
     * @return the parent span ID or null
     */
    public String getParentSpanId() {
        if (references == null)
            return null;

        for (Reference ref : references) {
            if (ref.isChildOf()) {
                return ref.getSpanID();
            }
        }
        return null;
    }

    /**
     * Checks if this is a root span (no parent).
     * 
     * @return true if this span has no parent
     */
    public boolean isRootSpan() {
        return getParentSpanId() == null;
    }

    @Override
    public String toString() {
        return "Span{" +
                "spanID='" + spanID + '\'' +
                ", operationName='" + operationName + '\'' +
                ", processID='" + processID + '\'' +
                ", duration=" + duration +
                '}';
    }
}

File: python_version/jaeger_uml_generator/models/span.py
"""Span model representing a traced operation."""

from dataclasses import dataclass, field
from typing import List, Dict, Any, Optional
from .reference import Reference


@dataclass
class Span:
    """Represents a single span in a Jaeger trace."""
    
    trace_id: str
    span_id: str
    operation_name: str
    start_time: int  # microseconds
    duration: int  # microseconds
    process_id: str
    references: List[Reference] = field(default_factory=list)
    tags: Dict[str, Any] = field(default_factory=dict)
    logs: List[Dict[str, Any]] = field(default_factory=list)
    
    @classmethod
    def from_dict(cls, data: dict) -> 'Span':
        """Create a Span from a dictionary."""
        # Parse references
        references = []
        if 'references' in data and isinstance(data['references'], list):
            references = [Reference.from_dict(ref) for ref in data['references']]
        
        # Parse tags
        tags = {}
        if 'tags' in data and isinstance(data['tags'], list):
            for tag in data['tags']:
                if isinstance(tag, dict) and 'key' in tag:
                    tags[tag['key']] = tag.get('value')
        
        # Parse logs
        logs = data.get('logs', [])
        
        return cls(
            trace_id=data.get('traceID', ''),
            span_id=data.get('spanID', ''),
            operation_name=data.get('operationName', ''),
            start_time=data.get('startTime', 0),
            duration=data.get('duration', 0),
            process_id=data.get('processID', ''),
            references=references,
            tags=tags,
            logs=logs
        )
    
    def get_parent_span_id(self) -> Optional[str]:
        """Get the parent span ID if this span has a parent."""
        for ref in self.references:
            if ref.ref_type == 'CHILD_OF':
                return ref.span_id
        return None
    
    def is_root_span(self) -> bool:
        """Check if this is a root span (no parent)."""
        return self.get_parent_span_id() is None
    
    def get_tag(self, key: str, default: Any = None) -> Any:
        """Get a tag value by key."""
        return self.tags.get(key, default)
